### PR TITLE
Add Persistent Volume E2E in the context of a disrupted kubelet

### PR DESCRIPTION
**Pull request by copejon**
This PR adds a test suite for persistent volumes affected by a disrupted kubelet.  Two cases are presented:

1.  A volume mounted via PVC remains accessible after a kubelet restart.
2.  When a pod is deleted while the kubelet is down, the mounted volume is unmounted successfully.

### CODE REVIEW

**Comment by spxtr** [comment]
@k8s-bot ok to test

/release-note-none

File: test/e2e/persistent_volumes-disruptive.go [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"k8s.io/kubernetes/pkg/api/v1"
+	metav1 "k8s.io/kubernetes/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/pkg/client/clientset_generated/clientset"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	"strings"
```
Comment (jeffvance) [inline test/e2e/persistent_volumes-disruptive.go]: convention is to list all of the go packages together at the top

File: test/e2e/persistent_volumes-disruptive.go [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+		c = f.ClientSet
+		ns = f.Namespace.Name
+
+		// Start the NFS server pod.
+		if nfsServerPod == nil {
+			framework.Logf("[BeforeEach] Initializing NFS Server Pod")
+			nfsServerPod = startVolumeServer(c, nfsServerConfig)
+
+			framework.Logf("[BeforeEach] Configuring PersistentVolume")
+			nfsServerIP = nfsServerPod.Status.PodIP
```
Comment (jeffvance) [inline test/e2e/persistent_volumes-disruptive.go]: do you want an `Expect(len(nfsServerIP)).NotTo(BeZero())` ?

File: test/e2e/persistent_volumes-disruptive.go [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+			}
+		}
+		// Get the first ready node IP that is not hosting the NFS pod.
+		if clientNodeIP == "" {
+			framework.Logf("Designating test node")
+			nodes := framework.GetReadySchedulableNodesOrDie(c)
+			for _, node := range nodes.Items {
+				if node.Name != nfsServerPod.Spec.NodeName {
+					clientNode = &node
+					break
```
Comment (jeffvance) [inline test/e2e/persistent_volumes-disruptive.go]: shouldn't clientNodeIP be set here so that the next time the BeforeEach is called this step is skipped?

File: test/e2e/persistent_volumes-disruptive.go [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+	nodeIP, err := framework.GetHostExternalAddress(c, pod)
+	Expect(err).NotTo(HaveOccurred())
+	nodeIP = nodeIP + ":22"
+	sshResult, err := framework.SSH("/etc/init.d/kubelet "+string(kOp), nodeIP, framework.TestContext.Provider)
+	Expect(err).NotTo(HaveOccurred())
+	framework.LogSSHResult(sshResult)
+
+	// On restart, waiting for node NotReady prevents a race condition where the node takes a few moments to leave the
+	// Ready state which in turn short circuits WaitForNodeToBeReady()
+	if kOp == kStop || kOp == kRestart {
```
Comment (jeffvance) [inline test/e2e/persistent_volumes-disruptive.go]: why the test of `kOp` here an on L249 since the block is identical? Why not unconditionally call `WaitForNodeToBeNotReady()`?

File: test/e2e/persistent_volumes-disruptive.go [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+	})
+})
+
+// SPEC: Test that a volume mounted to a pod remains mounted after a kubelet restarts
+func testKubeletRestartsAndRestoresMount(c clientset.Interface, f *framework.Framework, clientPod *v1.Pod, pvc *v1.PersistentVolumeClaim, pv *v1.PersistentVolume) {
+	file := "/mnt/_SUCCESS"
+	_, err := podExec(clientPod, "touch "+file)
+	Expect(err).NotTo(HaveOccurred())
+
+	kubeletCommand(kRestart, c, clientPod)
```
Comment (jeffvance) [inline test/e2e/persistent_volumes-disruptive.go]: Is `clientPod` stale when this func returns, meaning a different UID? I guess the fields your referencing below are still valid...

File: test/e2e/persistent_volumes-disruptive.go [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+	_, err = podExec(clientPod, "cat "+file)
+	Expect(err).NotTo(HaveOccurred())
+	framework.Logf("Pod %s detected %s after kubelet restart", clientPod.Name, file)
+}
+
+// SPEC: Test that a volume unmounts if the client pod was deleted while the kubelet was down.
+func testVolumeUnmountsFromDeletedPod(c clientset.Interface, f *framework.Framework, clientPod *v1.Pod, pvc *v1.PersistentVolumeClaim, pv *v1.PersistentVolume) {
+
+	nodeIP, err := framework.GetHostExternalAddress(c, clientPod)
+	nodeIP = nodeIP + ":22"
```
Comment (jeffvance) [inline test/e2e/persistent_volumes-disruptive.go]: nit: more typical to see the assignment after the Expect()

File: test/e2e/persistent_volumes-disruptive.go [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+		)
+
+		BeforeEach(func() {
+			framework.Logf("Initializing test spec")
+			clientPod, pv, pvc = initTestCase(f, c, nfsPVconfig, ns, clientNode.Name)
+		})
+
+		AfterEach(func() {
+			framework.Logf("Tearing down test spec")
+			tearDownTestCase(c, f, ns, clientPod, pvc, pv)
```
Comment (jeffvance) [inline test/e2e/persistent_volumes-disruptive.go]: is there any need to set clientPod, pv, pvc to nil after this call?

File: test/e2e/persistent_volumes-disruptive.go [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+		if ok := framework.WaitForNodeToBeReady(c, pod.Spec.NodeName, NODE_STATE_WAIT); !ok {
+			framework.Failf("Node %s failed to enter NotReady state", pod.Spec.NodeName)
+		}
+	}
+
+}
+
+// Wraps RunKubectl to execute a bash cmd in target pod
+func podExec(pod *v1.Pod, bashExec string) (string, error) {
+	args := strings.Split(bashExec, " ")
```
Comment (jeffvance) [inline test/e2e/persistent_volumes-disruptive.go]: why is this needed?

File: test/e2e/persistent_volumes-disruptive.go [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+		nfsServerPod              *v1.Pod
+		nfsPVconfig               persistentVolumeConfig
+		nfsServerIP, clientNodeIP string
+		clientNode                *v1.Node
+	)
+
+	nfsServerConfig := VolumeTestConfig{
+		namespace:   v1.NamespaceDefault,
+		prefix:      "nfs",
+		serverImage: "gcr.io/google_containers/volume-nfs:0.7",
```
Comment (copejon) [inline test/e2e/persistent_volumes-disruptive.go]: That's a good question.  Historically, it's the image we've been using for all persistent volume testing.  There is a `volume-nfs:0.8` used in `volumes.go` and `util.go`.  Looking at `test/images/volumes-tester/nfs/Makefile`, that appears to be the latest iteration.

Comment (jeffvance) [inline test/e2e/persistent_volumes-disruptive.go]: And there's some discussion related to mount,nfs support in the target os. See #33447

File: test/e2e/persistent_volumes-disruptive.go (hunk 5 above)
Comment (copejon) [inline test/e2e/persistent_volumes-disruptive.go]: I tested this manually to make sure and found that the pod UID remains the same through a kubelet restart.  Left out re-getting the pod to avoid unnecessary api calls.

File: test/e2e/persistent_volumes-disruptive.go (hunk 4 above)
Comment (copejon) [inline test/e2e/persistent_volumes-disruptive.go]: My intent here was to cut out an unnecessary api call. When `WaitForNodeToBeNotReady()` is called on `kubelet start`, it would make a Get() call and test for a NotReady state, which would be it's expected current state.
  
It also meant that the log would read like:

`ssh root@0.0.0.0: command:   /etc/init.d/kubelet start`
`Waiting up to 2m0s for node condition Ready to be false` 
`Condition Ready of node is true instead of false. Reason: KubeletReady, message: kubelet is posting ready status. WARNING: CPU hardcapping unsupported`
`Waiting up to 2m0s for node condition Ready to be true`
`Condition Ready of node jcope-e2e-minion-group-fmhy is false instead of true.`

Which seemed like it might be a little confusing.

I could be over thinking this a bit though.  Maybe the above reasons are a bit trivial.

File: test/e2e/persistent_volumes-disruptive.go (hunk 5 above)
Comment (jeffvance) [inline test/e2e/persistent_volumes-disruptive.go]: ok, that interesting and good to know!

File: test/e2e/persistent_volumes-disruptive.go [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+}
+
+// SPEC: Test that a volume unmounts if the client pod was deleted while the kubelet was down.
+func testVolumeUnmountsFromDeletedPod(c clientset.Interface, f *framework.Framework, clientPod *v1.Pod, pvc *v1.PersistentVolumeClaim, pv *v1.PersistentVolume) {
+
+	nodeIP, err := framework.GetHostExternalAddress(c, clientPod)
+	nodeIP = nodeIP + ":22"
+	Expect(err).NotTo(HaveOccurred())
+
+	result, err := framework.SSH("mount", nodeIP, framework.TestContext.Provider)
```
Comment (copejon) [inline test/e2e/persistent_volumes-disruptive.go]: Yeah, I like that more. Then you can `Expect(result.Code).To(BeZero())` || `.NotTo(BeZero())`

File: test/e2e/persistent_volumes-disruptive.go (hunk 4 above)
Comment (jeffvance) [inline test/e2e/persistent_volumes-disruptive.go]: Oh, never mind, my mistake. I read too quickly and thought that you called `WaitForNodeToBeNotReady` in both blocks.

File: test/e2e/persistent_volumes-disruptive.go [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+	} else {
+		framework.Logf("Client UID %s not found in `mount` output.  Volume has unmounted.", clientPod.UID)
+	}
+}
+
+// Initialize a test spec with a pv, pvc, and nfs client pod
+func initTestCase(f *framework.Framework, c clientset.Interface, pvConfig persistentVolumeConfig, ns, nodeName string) (pod *v1.Pod, pv *v1.PersistentVolume, pvc *v1.PersistentVolumeClaim) {
+	pv, pvc = createPVPVC(c, pvConfig, ns, false)
+	pod = makePod(ns, pvc.Name)
+	pod.Spec.NodeName = nodeName
```
Comment (copejon) [inline test/e2e/persistent_volumes-disruptive.go]: From the v1 Pod.Spec [api definitions](http://kubernetes.io/docs/api-reference/v1/definitions/):
> NodeName is a request to schedule this pod onto a specific node. If it is non-empty, the scheduler simply schedules this pod onto that node, assuming that it fits resource requirements.

Using nodeLabels/Selectors is another option but a felt like overkill for the test.

Comment (jeffvance) [inline test/e2e/persistent_volumes-disruptive.go]: ok, I wondered if that was the reason, but I thought label/selectors would be used for that. But I agree with your comment: it is simpler.

File: test/e2e/persistent_volumes-disruptive.go (hunk 8 above)
Comment (copejon) [inline test/e2e/persistent_volumes-disruptive.go]: Just to keep the thread up w/ our offline convo:
`framework.RunKubectl(cmd...)` fails w/ 
>       Command stdout:
>        rpc error: code = 2 desc = "oci runtime error: exec failed: exec: \"touch /mnt/_SUCCESS\": stat touch /mnt/_SUCCESS: no such file or directory"
 
when the bash command string isn't broken into separate strings.

File: test/e2e/persistent_volumes-disruptive.go (hunk 9 above)
Comment (jsafrane) [inline test/e2e/persistent_volumes-disruptive.go]: @copejon, can I ask you to create a const somewhere with the latest "gcr.io/google_containers/volume-nfs:XXX" and use it in all tests that need it? Or even export const with whole `VolumeTestConfig`, if it is the same everywhere.

File: test/e2e/persistent_volumes-disruptive.go [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+	Expect(err).NotTo(HaveOccurred())
+	if strings.Contains(result.Stdout, string(clientPod.UID)) {
+		framework.Logf("Sanity Check: Client UID %s found in `mount` output", clientPod.UID)
+	} else {
+		framework.Failf("Sanity Check: Client UID %s NOT found in `mount` output prior to pod deletion.  Something has gone wrong.", clientPod.UID)
+	}
+
+	file := "/mnt/_SUCCESS"
+	_, err = podExec(clientPod, "touch "+file)
+	Expect(err).NotTo(HaveOccurred())
```
Comment (jsafrane) [inline test/e2e/persistent_volumes-disruptive.go]: You write _SUCCESS to the volume but you never check that it's gone after unmount.

File: test/e2e/persistent_volumes-disruptive.go (hunk 11 above)
Comment (jsafrane) [inline test/e2e/persistent_volumes-disruptive.go]: yes, setting `pod.Spec.NodeName` is the right thing to do, we want to bypass scheduler here.

File: test/e2e/persistent_volumes-disruptive.go [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+	nodeIP = nodeIP + ":22"
+	sshResult, err := framework.SSH("/etc/init.d/kubelet "+string(kOp), nodeIP, framework.TestContext.Provider)
+	Expect(err).NotTo(HaveOccurred())
+	framework.LogSSHResult(sshResult)
+
+	// On restart, waiting for node NotReady prevents a race condition where the node takes a few moments to leave the
+	// Ready state which in turn short circuits WaitForNodeToBeReady()
+	if kOp == kStop || kOp == kRestart {
+		if ok := framework.WaitForNodeToBeNotReady(c, pod.Spec.NodeName, NODE_STATE_WAIT); !ok {
```
Comment (jsafrane) [inline test/e2e/persistent_volumes-disruptive.go]: with `kRestart`, can be the test too slow and miss "NotReady" because the node was too quick and was down for a very short time so the test did not notice?

File: test/e2e/persistent_volumes-disruptive.go [hunk 14]
```diff
@@ -0,0 +1,263 @@
...
+				It(t.testItStmt+" [Disruptive]", func() {
+					By("Executing Spec")
+					t.runTest(c, f, clientPod, pvc, pv)
+				})
+			}(test)
+		}
+	})
+})
+
+// SPEC: Test that a volume mounted to a pod remains mounted after a kubelet restarts
```
Comment (jsafrane) [inline test/e2e/persistent_volumes-disruptive.go]: by convention, function documentation starts with the function name, e.g. "testKubeletRestartsAndRestoresMount checks that a volume mounted ...". This applies to documentation of all functions in this PR.

File: test/e2e/persistent_volumes-disruptive.go (hunk 9 above)
Comment (copejon) [inline test/e2e/persistent_volumes-disruptive.go]: @jsafrane will do, I'll place it with `VolumeTestConfig`.

Comment (copejon) [inline test/e2e/persistent_volumes-disruptive.go]: @jsafrane Okay.  Since there's also gcr.io images for glusterfs, rbd, ceph, and iscsi,  I think a volumeTestConfig constructor would the route to take.  As a param, it can take a `const <image path> string` along with other relevant fields.  Will work that out and push it.

File: test/e2e/persistent_volumes-disruptive.go (hunk 13 above)
Comment (jsafrane) [inline test/e2e/persistent_volumes-disruptive.go]: I just got this idea: you can stop the service, wait until it's NotReady and then start it, this way it's bulletproof.

**Comment by jsafrane** [comment]
Thanks for the PR!

/lgtm
